base/services/w32time/W32Time.c

```diff
@@ -0,0 +1,366 @@
+                                          NULL,
+                                          0);
+                }
+            }
+        }
+        CloseHandle(hToken);
+    }
+
+    return Ret;
+}
+
+static DWORD
+GetIntervalSetting()
+{
+    HKEY hKey;
+    DWORD dwData;
+    DWORD dwSize = sizeof(DWORD);
+
+    dwData = 0;
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                      L"SYSTEM\\CurrentControlSet\\Services\\W32Time\\TimeProviders\\NtpClient",
+                      0,
+                      KEY_QUERY_VALUE,
+                      &hKey) == ERROR_SUCCESS)
+    {
+        RegQueryValueExW(hKey,
+                             L"SpecialPollInterval",
+                             NULL,
+                             NULL,
+                             (LPBYTE)&dwData,
+                             &dwSize);
+        RegCloseKey(hKey);
+    }
+    return dwData;
+}
+
+
+int
+SetTime(VOID)
+{
+    ULONG ulTime;
+    LONG lRet;
+    HKEY hKey;
+    DWORD dwIndex = 0;
+    DWORD dwValSize;
+    DWORD dwNameSize;
+    WCHAR szValName[MAX_VALUE_NAME];
+    WCHAR szData[256];
+    WCHAR szDefault[256];
+
+
+    lRet = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                        L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\DateTime\\Servers",
+                        0,
+                        KEY_QUERY_VALUE,
+                        &hKey);
+    if (lRet != ERROR_SUCCESS)
+    {
+        return 1;
+    }
+
+    while (TRUE)
+    {
+        dwValSize = MAX_VALUE_NAME * sizeof(WCHAR);
+        szValName[0] = L'\0';
+        lRet = RegEnumValueW(hKey,
+                             dwIndex,
+                             szValName,
+                             &dwValSize,
+                             NULL,
+                             NULL,
+                             (LPBYTE)szData,
+                             &dwNameSize);
+        if (lRet == ERROR_SUCCESS)
+        {
+            /* Get data from default reg value */
+            if (szValName[0] == UNICODE_NULL) // if this is the "(Default)" key
+            {
+                StringCbCopy(szDefault, wcslen(szDefault), szData);
+            }
+            dwIndex++;
+        }
+        else if (lRet != ERROR_MORE_DATA)
+        {
+            break;
+        }
+    }
+
+
+    dwIndex = 0;
+    while (TRUE)
+    {
+        dwValSize = MAX_VALUE_NAME * sizeof(WCHAR);
+        szValName[0] = L'\0';
+        lRet = RegEnumValueW(hKey,
+                             dwIndex,
+                             szValName,
+                             &dwValSize,
+                             NULL,
+                             NULL,
+                             (LPBYTE)szData,
+                             &dwNameSize);
+        if (lRet == ERROR_SUCCESS)
+        {
+            /* Get date from selected reg value */
+            if (wcscmp(szValName, szDefault) == 0) // if (Index == Default)
+            {
+                StringCbCopy(szDefault, wcslen(szDefault), szData);
```

Comment:
Check my remark for this function call above.
Also, please ensure that szValName is NULL-terminated before calling wcscmp.

Reply:
I eliminated the wcscmp and changed this line to "if (szValName[0] == UNICODE_NULL)". Thanks.